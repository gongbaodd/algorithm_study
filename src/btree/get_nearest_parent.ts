import { BTree } from "./lib";
import { Stack } from "../stack/lib";

function getPathFromRoot<T>(
  root: BTree<T> | null,
  node: typeof root,
  stack: Stack<typeof root>
): boolean {
  if (root === null) {
    return false;
  }

  if (node === root) {
    stack.push(root);
    return true;
  }

  if (
    getPathFromRoot(root.lChild, node, stack) ||
    getPathFromRoot(root.rChild, node, stack)
  ) {
    stack.push(root);
    return true;
  }

  return false;
}

export function findParentNode<T>(
  root: BTree<T> | null,
  node1: typeof root,
  node2: typeof root
): typeof root {
  const stack1 = new Stack<typeof root>();
  const stack2 = new Stack<typeof root>();

  getPathFromRoot(root, node1, stack1);
  getPathFromRoot(root, node2, stack2);

  let parent: typeof root = null;

  while (stack1.top() === stack2.top()) {
    parent = stack1.top();
    stack1.pop();
    stack2.pop();
  }

  return parent;
}
