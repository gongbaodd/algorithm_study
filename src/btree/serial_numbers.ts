import { BTree } from "./lib";

export function toTree(
  arr: number[],
  start: number,
  end: number
): BTree<number> | null {
  let root: BTree<number> | null = new BTree<number>();

  if (end < start) {
    root = null;
    return root;
  }

  const mid = parseInt(((start + end) / 2).toString(), 10);

  root.data = arr[mid];

  root.lChild = toTree(arr, start, mid - 1);
  root.rChild = toTree(arr, mid + 1, end);

  return root;
}
