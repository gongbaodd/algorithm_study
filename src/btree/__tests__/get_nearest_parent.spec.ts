import { findParentNode } from "../get_nearest_parent";
import { toTree } from "../serial_numbers";

describe("find nearest  parent", () => {
  it("should find the parent", () => {
    const root = toTree([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 0, 9);

    const node1 = root?.lChild?.lChild ?? null;
    const node2 = root?.lChild?.rChild ?? null;

    const parent = findParentNode(root, node1, node2);

    expect(parent?.data).toBe(2);
    expect(parent).toMatchSnapshot();
  });
});
