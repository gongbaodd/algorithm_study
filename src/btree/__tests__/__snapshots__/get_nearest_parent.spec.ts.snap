// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find nearest  parent should find the parent 1`] = `
BTree {
  "data": 2,
  "lChild": BTree {
    "data": 1,
    "lChild": null,
    "rChild": null,
  },
  "rChild": BTree {
    "data": 3,
    "lChild": null,
    "rChild": BTree {
      "data": 4,
      "lChild": null,
      "rChild": null,
    },
  },
}
`;
